File: scripts/genMap.go
package main

import (
	"fmt"
	"math/rand"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"strconv"
	"strings"
	"sync"
	"time"
)

// Cell represents a character in the map grid.
type Cell rune

const (
	Wall    Cell = '1'
	Floor   Cell = '0'
	North   Cell = 'N'
	South   Cell = 'S'
	East    Cell = 'E'
	West    Cell = 'W'
	Space   Cell = ' '
)

// Random utility functions
func randInt(max int) int {
	return rand.Intn(max)
}

func choose[T any](arr []T) T {
	return arr[randInt(len(arr))]
}

func randomRGB() (r, g, b int) {
	return randInt(256), randInt(256), randInt(256)
}

// Grid operations
func makeEmptyGrid(w, h int) [][]Cell {
	grid := make([][]Cell, h)
	for y := 0; y < h; y++ {
		grid[y] = make([]Cell, w)
		for x := 0; x < w; x++ {
			grid[y][x] = Wall
		}
	}
	return grid
}

func gridToLines(grid [][]Cell) []string {
	lines := make([]string, len(grid))
	for i, row := range grid {
		var sb strings.Builder
		for _, cell := range row {
			sb.WriteRune(rune(cell))
		}
		lines[i] = sb.String()
	}
	return lines
}

func placeInteriorRandom(grid [][]Cell, floorProb float64) {
	h := len(grid)
	if h == 0 {
		return
	}
	w := len(grid[0])
	for y := 1; y < h-1; y++ {
		for x := 1; x < w-1; x++ {
			if rand.Float64() < floorProb {
				grid[y][x] = Floor
			} else {
				grid[y][x] = Wall
			}
		}
	}
}

func findAll(grid [][]Cell, chs map[Cell]bool) [][2]int {
	var out [][2]int
	for y := 0; y < len(grid); y++ {
		for x := 0; x < len(grid[0]); x++ {
			if chs[grid[y][x]] {
				out = append(out, [2]int{x, y})
			}
		}
	}
	return out
}

func countPlayer(grid [][]Cell) int {
	playerChars := map[Cell]bool{North: true, South: true, East: true, West: true}
	return len(findAll(grid, playerChars))
}

func placePlayer(grid [][]Cell) bool {
	floorChars := map[Cell]bool{Floor: true}
	floors := findAll(grid, floorChars)
	if len(floors) == 0 {
		return false
	}
	pos := choose(floors)
	x, y := pos[0], pos[1]
	dir := choose([]Cell{North, South, East, West})
	grid[y][x] = dir
	return true
}

func hasInvalidChars(grid [][]Cell) bool {
	allowed := map[Cell]bool{Floor: true, Wall: true, North: true, South: true, East: true, West: true, Space: true}
	for _, row := range grid {
		for _, c := range row {
			if !allowed[c] {
				return true
			}
		}
	}
	return false
}

// Core Validation Function (flood-fill)
func floodFillIsValid(grid [][]Cell) bool {
	h := len(grid)
	if h == 0 {
		return false
	}
	w := len(grid[0])
	if w == 0 {
		return false
	}

	// 1. Check for exactly one player & no spaces in map body
	px, py, pcnt := -1, -1, 0
	for y := 0; y < h; y++ {
		for x := 0; x < w; x++ {
			c := grid[y][x]
			if c == North || c == South || c == East || c == West {
				px, py = x, y
				pcnt++
			}
			if c == Space {
				// Spaces are invalid in the main map region for this validation style
				return false
			}
		}
	}
	if pcnt != 1 {
		return false
	}

	// 2. BFS/DFS flood-fill from player
	visited := make([][]bool, h)
	for i := range visited {
		visited[i] = make([]bool, w)
	}

	queue := [][2]int{{px, py}}
	visited[py][px] = true

	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]
		x, y := curr[0], curr[1]

		// If we are at border (0 or w-1 / h-1) => invalid (unclosed map)
		if x == 0 || y == 0 || x == w-1 || y == h-1 {
			return false
		}

		neighbors := [][2]int{
			{x + 1, y}, {x - 1, y}, {x, y + 1}, {x, y - 1},
		}

		for _, nei := range neighbors {
			nx, ny := nei[0], nei[1]

			if nx < 0 || ny < 0 || nx >= w || ny >= h {
				continue // Should not happen with border check above, but safe to include
			}
			if visited[ny][nx] {
				continue
			}

			c := grid[ny][nx]

			if c == Wall {
				continue // Wall, stop here
			}
			if c == Space {
				return false // Invalid space encountered
			}

			visited[ny][nx] = true
			queue = append(queue, nei)
		}
	}
	return true
}

// File handling and Map Generation Orchestration
func writeMapFile(path string, headers []string, lines []string) error {
	content := strings.Join(headers, "\n") + "\n\n" + strings.Join(lines, "\n") + "\n"
	return os.WriteFile(path, []byte(content), 0644)
}

func parseOutputPaths(outPath string) ([]string, error) {
	rangeRe := regexp.MustCompile(`\[(\d+)\.\.(\d+)\]`)
	m := rangeRe.FindStringSubmatch(outPath)
	if m == nil {
		return []string{outPath}, nil
	}

	startStr, endStr := m[1], m[2]
	start, errStart := strconv.Atoi(startStr)
	end, errEnd := strconv.Atoi(endStr)
	if errStart != nil || errEnd != nil {
		return nil, fmt.Errorf("invalid range numbers in out_path: %v, %v", errStart, errEnd)
	}

	if start > end {
		start, end = end, start // Normalize direction
	}

	var outputs []string
	pad := 0
	if len(startStr) > 1 {
		pad = len(startStr)
	}

	for i := start; i <= end; i++ {
		idx := strconv.Itoa(i)
		if pad > 0 {
			idx = fmt.Sprintf("%0*d", pad, i)
		}
		outputs = append(outputs, rangeRe.ReplaceAllString(outPath, idx))
	}
	return outputs, nil
}

func generateMap(targetPath string, w, h, maxAttempts int, wg *sync.WaitGroup) {
	defer wg.Done()

	var (
		attempt = 0
		success = false
		grid    [][]Cell
	)

	for attempt < maxAttempts && !success {
		attempt++
		grid = makeEmptyGrid(w, h)
		// Carve interior randomly (0.63 prob for floor)
		placeInteriorRandom(grid, 0.63)

		// Ensure at least one floor if none was carved
		floorChars := map[Cell]bool{Floor: true}
		if len(findAll(grid, floorChars)) == 0 {
			cx, cy := w/2, h/2
			grid[cy][cx] = Floor
		}

		// Clear existing player characters if any (shouldn't happen here, but robust)
		for y := 0; y < h; y++ {
			for x := 0; x < w; x++ {
				c := grid[y][x]
				if c == North || c == South || c == East || c == West {
					grid[y][x] = Floor
				}
			}
		}

		if !placePlayer(grid) {
			continue
		}

		// Quick checks (checking again after placePlayer to be safe, though placePlayer uses "0")
		if hasInvalidChars(grid) {
			continue
		}
		if countPlayer(grid) != 1 {
			continue
		}

		// Flood-fill validation (the critical check)
		if !floodFillIsValid(grid) {
			continue
		}
		success = true
	}

	if !success {
		fmt.Fprintf(os.Stderr, "Failed to generate a valid map after %d attempts for %s\n", maxAttempts, targetPath)
		return
	}

	// Generate map headers
	fr, fg, fb := randomRGB()
	cr, cg, cb := randomRGB()
	headers := []string{
		"NO ./north.xpm",
		"SO ./south.xpm",
		"WE ./west.xpm",
		"EA ./east.xpm",
		fmt.Sprintf("F %d,%d,%d", fr, fg, fb),
		fmt.Sprintf("C %d,%d,%d", cr, cg, cb),
	}

	lines := gridToLines(grid)

	// Ensure directory exists
	dir := filepath.Dir(targetPath)
	if dir != "" && dir != "." {
		if err := os.MkdirAll(dir, 0755); err != nil {
			fmt.Fprintf(os.Stderr, "Error creating directory for %s: %v\n", targetPath, err)
			return
		}
	}

	// Write file
	if err := writeMapFile(targetPath, headers, lines); err != nil {
		fmt.Fprintf(os.Stderr, "Error writing map file to %s: %v\n", targetPath, err)
		return
	}

	fmt.Printf("Wrote generated map to %s (attempt %d)\n", targetPath, attempt)
}

func main() {
	// Seed the random number generator
	rand.Seed(time.Now().UnixNano())

	// Use command-line arguments directly from os.Args for simplicity, similar to the Deno script
	if len(os.Args) < 3 {
		fmt.Fprintln(os.Stderr, "Usage: go run genMap.go <width> <height> [out_path]")
		os.Exit(1)
	}

	w, errW := strconv.Atoi(os.Args[1])
	h, errH := strconv.Atoi(os.Args[2])
	if errW != nil || errH != nil || w < 3 || h < 3 {
		fmt.Fprintln(os.Stderr, "Width and height must be integers >= 3")
		os.Exit(1)
	}

	outPath := fmt.Sprintf("assets/maps/generated_%dx%d.cub", w, h)
	if len(os.Args) >= 4 {
		outPath = os.Args[3]
	}

	outputs, err := parseOutputPaths(outPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error parsing output path: %v\n", err)
		os.Exit(1)
	}

	maxAttempts := 200

	// Use standard concurrency model (goroutines and WaitGroup)
	// We'll limit concurrency to the number of CPUs available.
	cpuCount := runtime.GOMAXPROCS(0)
	if cpuCount == 0 {
		cpuCount = 1
	}

	var wg sync.WaitGroup
	sem := make(chan struct{}, cpuCount) // Semaphore to limit concurrency

	for _, targetPath := range outputs {
		wg.Add(1)
		sem <- struct{}{} // Acquire a slot

		go func(path string) {
			defer func() { <-sem }() // Release the slot
			generateMap(path, w, h, maxAttempts, &wg)
		}(targetPath)
	}

	wg.Wait()
}

File: scripts/genMap.ts
#!/usr/bin/env -S deno run --allow-read --allow-write
// Generate a random but valid .cub map for the cub3d project.
// Usage: deno run --allow-read --allow-write scripts/genMap.ts <width> <height> [out_path]

function randInt(max: number) {
  return Math.floor(Math.random() * max);
}

function choose<T>(arr: T[]) {
  return arr[randInt(arr.length)];
}

function randomRGB(): [number, number, number] {
  return [randInt(256), randInt(256), randInt(256)];
}

type Cell = "0" | "1" | "N" | "S" | "E" | "W" | " ";

function makeEmptyGrid(w: number, h: number): Cell[][] {
  const grid: Cell[][] = new Array(h);
  for (let y = 0; y < h; y++) {
    grid[y] = new Array(w);
    for (let x = 0; x < w; x++) grid[y][x] = "1";
  }
  return grid;
}

function gridToLines(grid: Cell[][]): string[] {
  return grid.map((row) => row.join(""));
}

function placeInteriorRandom(grid: Cell[][], floorProb = 0.6) {
  const h = grid.length,
    w = grid[0].length;
  for (let y = 1; y < h - 1; y++) {
    for (let x = 1; x < w - 1; x++) {
      grid[y][x] = Math.random() < floorProb ? "0" : "1";
    }
  }
}

function findAll(grid: Cell[][], chs: Set<Cell>) {
  const out: [number, number][] = [];
  for (let y = 0; y < grid.length; y++) {
    for (let x = 0; x < grid[0].length; x++) {
      if (chs.has(grid[y][x])) out.push([x, y]);
    }
  }
  return out;
}

function countPlayer(grid: Cell[][]) {
  const set = new Set(["N", "S", "E", "W"]);
  return findAll(grid, set as Set<Cell>).length;
}

function placePlayer(grid: Cell[][]) {
  const floors = findAll(grid, new Set(["0"]));
  if (floors.length === 0) return false;
  const [x, y] = choose(floors);
  const dir = choose(["N", "S", "E", "W"]) as Cell;
  grid[y][x] = dir;
  return true;
}

function hasInvalidChars(grid: Cell[][]) {
  const allowed = new Set(["0", "1", "N", "S", "E", "W", " "]);
  for (const row of grid) for (const c of row) if (!allowed.has(c)) return true;
  return false;
}

function floodFillIsValid(grid: Cell[][]): boolean {
  const h = grid.length,
    w = grid[0].length;
  // Check for exactly one player
  let px = -1,
    py = -1,
    pcnt = 0;
  for (let y = 0; y < h; y++) {
    for (let x = 0; x < w; x++) {
      const c = grid[y][x];
      if (c === "N" || c === "S" || c === "E" || c === "W") {
        px = x;
        py = y;
        pcnt++;
      }
      if (c === " ") return false; // spaces are treated as invalid region per validator
    }
  }
  if (pcnt !== 1) return false;

  // BFS/DFS flood-fill from player; cannot reach boundary or space
  const visited = Array.from({ length: h }, () =>
    new Array<boolean>(w).fill(false),
  );
  const q: [number, number][] = [];
  q.push([px, py]);
  visited[py][px] = true;
  while (q.length > 0) {
    const [x, y] = q.shift()!;
    // If we are at border => invalid
    if (x === 0 || y === 0 || x === w - 1 || y === h - 1) return false;
    const nei = [
      [x + 1, y],
      [x - 1, y],
      [x, y + 1],
      [x, y - 1],
    ];
    for (const [nx, ny] of nei) {
      if (nx < 0 || ny < 0 || nx >= w || ny >= h) continue;
      if (visited[ny][nx]) continue;
      const c = grid[ny][nx];
      if (c === "1") continue; // wall
      if (c === " ") return false; // invalid
      visited[ny][nx] = true;
      q.push([nx, ny]);
    }
  }
  return true;
}

async function writeMapFile(path: string, headers: string[], lines: string[]) {
  const content = headers.join("\n") + "\n\n" + lines.join("\n") + "\n";
  await Deno.writeTextFile(path, content);
}

async function main() {
  const args = Deno.args;
  if (args.length < 2) {
    console.error("Usage: genMap.ts <width> <height> [out_path]");
    Deno.exit(1);
  }
  const w = Number(args[0]);
  const h = Number(args[1]);
  if (!Number.isInteger(w) || !Number.isInteger(h) || w < 3 || h < 3) {
    console.error("Width and height must be integers >= 3");
    Deno.exit(1);
  }
  let outPath = args[2] ?? `assets/maps/generated_${w}x${h}.cub`;

  // Support smart range syntax in outPath, e.g. "maps/map[0..2].cub" -> 3 maps
  const rangeRe = /\[(\d+)\.\.(\d+)\]/;
  const outputs: string[] = [];
  const m = outPath.match(rangeRe);
  if (m) {
    const startStr = m[1];
    const endStr = m[2];
    let start = Number(startStr);
    let end = Number(endStr);
    if (Number.isNaN(start) || Number.isNaN(end)) {
      console.error("Invalid range in out_path");
      Deno.exit(1);
    }
    // allow either direction by normalizing
    if (start > end) {
      const tmp = start;
      start = end;
      end = tmp;
    }
    const pad = startStr.length > 1 ? startStr.length : 0;
    for (let i = start; i <= end; i++) {
      const idx = pad ? String(i).padStart(pad, "0") : String(i);
      outputs.push(outPath.replace(rangeRe, idx));
    }
  } else {
    outputs.push(outPath);
  }

  const maxAttempts = 200;

  // Generate function for a single map
  async function generateMap(targetPath: string): Promise<void> {
    let attempt = 0;
    let success = false;
    let grid: Cell[][] = [];

    while (attempt < maxAttempts && !success) {
      attempt++;
      grid = makeEmptyGrid(w, h);
      // Carve interior randomly
      placeInteriorRandom(grid, 0.63);
      // Ensure at least one floor
      const floors = findAll(grid, new Set(["0"]));
      if (floors.length === 0) {
        const cx = Math.floor(w / 2),
          cy = Math.floor(h / 2);
        grid[cy][cx] = "0";
      }
      // Place player
      // clear any existing players just in case
      for (let y = 0; y < h; y++) {
        for (let x = 0; x < w; x++) {
          if (["N", "S", "E", "W"].includes(grid[y][x])) grid[y][x] = "0";
        }
      }
      if (!placePlayer(grid)) continue;
      // quick checks
      if (hasInvalidChars(grid)) continue;
      if (countPlayer(grid) !== 1) continue;
      // flood-fill validation
      if (!floodFillIsValid(grid)) continue;
      success = true;
    }

    if (!success) {
      console.error(
        "Failed to generate a valid map after",
        maxAttempts,
        "attempts for",
        targetPath,
      );
      return;
    }

    const [fr, fg, fb] = randomRGB();
    const [cr, cg, cb] = randomRGB();
    const headers = [
      "NO ./north.xpm",
      "SO ./south.xpm",
      "WE ./west.xpm",
      "EA ./east.xpm",
      `F ${fr},${fg},${fb}`,
      `C ${cr},${cg},${cb}`,
    ];

    const lines = gridToLines(grid);
    // Ensure directory exists for targetPath
    try {
      const dir = targetPath.substring(0, targetPath.lastIndexOf("/")) || ".";
      await Deno.mkdir(dir, { recursive: true });
    } catch (_e) {}
    await writeMapFile(targetPath, headers, lines);
    console.log(`Wrote generated map to ${targetPath} (attempt ${attempt})`);
  }

  // Generate maps with concurrency limited to CPU count
  const cpuCount = navigator.hardwareConcurrency || 1;
  const results: Promise<void>[] = [];

  for (let i = 0; i < outputs.length; i += cpuCount) {
    const batch = outputs.slice(i, i + cpuCount);
    await Promise.all(batch.map(generateMap));
  }
}

if (import.meta.main) {
  main();
}
